### PR TITLE
Enhance RenameMap to support circuit renaming

### PR DESCRIPTION
**Pull request by jackkoenig**
Also dynamically enforce some type safety

### CODE REVIEW

**Comment by jackkoenig** [comment]
The fails are spurious, but unfortunately this change does seem to have a pretty large performance impact. It makes sense, every single ComponentName lookup requires 3 gets from the map, not to mention function call overhead. I'm looking into ways to measuring the performance and improving it.

**Comment by jackkoenig** [comment]
After further evaluation, the performance impact is actually quite small (my test case was to blame for how bad I perceived the impact to be). I think this PR is good and ready to merge.

**Comment by colinschmidt** [comment]
Cool. I wanted this along time ago and couldn't even work around it within the rename map see https://github.com/ucb-bar/barstools/commit/70805f137c3f2403b8f34530dbca26cba72bf37b

File: src/main/scala/firrtl/Compiler.scala [hunk 1]
```diff
@@ -27,42 +27,61 @@ object RenameMap {
...
-    * This is analogous to get on standard Scala collection Maps
-    * None indicates the key was not renamed
-    * Empty indicates the name was deleted
+
+  /** Get renames of a [[CircuitName]]
+    * @note A [[CircuitName]] can only be renamed to one-or-more [[CircuitName]]s
     */
-  // TODO Is there a better way to express this?
-  def get(key: Named): Option[Seq[Named]] = {
+  def get(key: CircuitName): Option[Seq[CircuitName]] = underlying.get(key).map {
```
Comment (colinschmidt) [inline src/main/scala/firrtl/Compiler.scala]: How do you rename it to multiple `CircuitName`s? Is that splitting the circuit into two circuits? Does the rest of firrtl support that?

File: src/main/scala/firrtl/Compiler.scala [hunk 2]
```diff
@@ -27,42 +27,61 @@ object RenameMap {
...
+    names => names.map {
+      case c: CircuitName => c
+      case other => error(s"Unsupported rename of $key to $other!")
```
Comment (colinschmidt) [inline src/main/scala/firrtl/Compiler.scala]: Maybe say `Unsupported Circuit rename...`

File: src/test/scala/firrtlTests/RenameMapSpec.scala [hunk 3]
```diff
@@ -71,4 +75,55 @@ class RenameMapSpec extends FirrtlFlatSpec {
     renames.rename(foo, bar)
     renames.get(foo) should be (Some(Seq(barB)))
   }
+
+  it should "rename modules if their circuit is renamed" in {
+    val renames = RenameMap()
+    renames.rename(cir, cir2)
+    renames.get(modA) should be (Some(Seq(modA2)))
```
Comment (colinschmidt) [inline src/test/scala/firrtlTests/RenameMapSpec.scala]: Shouldn't you also check that modB renamed?

File: src/test/scala/firrtlTests/RenameMapSpec.scala [hunk 4]
```diff
@@ -71,4 +75,55 @@ class RenameMapSpec extends FirrtlFlatSpec {
...
+  }
+
+  it should "rename components if their circuit is renamed" in {
+    val renames = RenameMap()
+    renames.rename(cir, cir2)
+    renames.get(foo) should be (Some(Seq(foo2)))
```
Comment (colinschmidt) [inline src/test/scala/firrtlTests/RenameMapSpec.scala]: Can we also check fooB for renaming

File: src/test/scala/firrtlTests/RenameMapSpec.scala (hunk 3 above)
Comment (jackkoenig) [reply src/test/scala/firrtlTests/RenameMapSpec.scala]: In this case, the only thing in the rename map is the circuit rename. Thus checking for modA and modB is identical.

File: src/test/scala/firrtlTests/RenameMapSpec.scala (hunk 4 above)
Comment (jackkoenig) [reply src/test/scala/firrtlTests/RenameMapSpec.scala]: Same as above

File: src/test/scala/firrtlTests/RenameMapSpec.scala [hunk 5]
```diff
@@ -88,42 +90,27 @@ class RenameMapSpec extends FirrtlFlatSpec {
     renames.get(foo) should be (Some(Seq(foo2)))
   }
 
-  it should "error if a component is renamed to a module or circuit" in {
-    a [FIRRTLException] shouldBe thrownBy {
+  // Renaming from to all of the tos at once should error
```
Comment (colinschmidt) [inline src/test/scala/firrtlTests/RenameMapSpec.scala]: Renaming from to each of the tos one at a time should error

**Pull request merged**